sympy/parsing/sympy_parser.py

```diff
@@ -507,6 +509,95 @@ def implicit_multiplication_application(result, local_dict, global_dict):
 
     return result
 
+def find_closing(string,pos,opening,closing):
+    # find the next closing
+    ic=string.index(closing,pos)
+    # print("ic",ic)
+    # find the next opening in the space between if there is one
+    try:
+        io=string.index(opening,pos+1,ic-1)
+        # print("io",io)
+        # find the closing to the next (leftmost) opening
+        ioc=find_closing(string,io,opening,closing)
+        # and start from there
+        return string.index(closing,ioc+1)
+    except ValueError:
+        #if there is no opening the next closing was the matching one
+        return ic
+
+
+def auto_indexed_base(tokens, local_dict, global_dict):
+    """Inserts calls to ``IndexedBase`` for undefined indexed variables."""
+    result = []
+    prevTok = (None, None)
+    n=len(tokens)
+    local_dict=copy.copy(local_dict)
+    tokens.append((None, None))  # to provide a last nextTok
```

Comment:
You can also make use of the fact that the last token emitted by tokenize is always `ENDMARKER`.